xontrib/hist_navigator.py

```diff
@@ -90,24 +126,114 @@ def insert_text(event, text):
     carriage_return(b, event.cli)
 
 
+def _cd_inline(path: typing.Optional[typing.AnyStr] = None) -> None:
+    """Change dir without creating a new prompt line, updating existing instead"""
+
+    if path is None:
+        args = []
+    elif isinstance(path, bytes):
+        args = [path.decode("utf-8")]
+    elif isinstance(path, str):
+        args = [path]
+    if         path and \
+      not Path(path).is_dir():
+        return # do nothing is target is not a Dir
+    _, exc, _ = cdx(args)
+    if exc is not None:
+        raise Exception(exc)
+    else:
+        t = threading.Thread(target=_update_prompt, args=())
+        t.start()
+
+
 @XSH.builtins.events.on_ptk_create  # noqa
 def custom_keybindings(bindings, **_):
-    handler = bindings.add
-
-    @handler("escape", "left", filter=cmd_empty_prompt)
+    from prompt_toolkit.key_binding.key_bindings import _parse_key
+
+    re_despace = re.compile(r'\s', re.IGNORECASE)
+
+    _default_keys = {
+        "X_HISTNAV_KEY_PREV" : ['escape','left' ],
+        "X_HISTNAV_KEY_NEXT" : ['escape','right'],
+        "X_HISTNAV_KEY_UP"   : ['escape','up'   ],
+    }
+
+    _key_symb = {
+        '⎈':'c-'  ,'⌃':'c-'   ,
+        '▼':'down' ,'↓':'down' ,
+        '▲':'up'   ,'↑':'up'   ,
+        '◀':'left' ,'←':'left' ,
+        '▶':'right','→':'right',
+    }
+    _alts = ['a-','⌥','⎇']
+
+    def handler(key_user_var, filter):
+        def skip(func):
+          pass
+
+        if envx.get('SHELL_TYPE') in ["prompt_toolkit", "prompt_toolkit2"]:
```

Comment:
if possible could you split this into small functions.

Reply:
moved is_ptk check and key parser out